src/blob/index/bptree/serializer.rs

```diff
@@ -0,0 +1,269 @@
+use super::prelude::*;
+
+struct HeaderStage {
+    header: IndexHeader,
+    key_size: usize,
+    filter_buf: Vec<u8>,
+}
+
+struct LeavesStage {
+    leaf_size: usize,
+    leaves_buf: Vec<u8>,
+    leaves_offset: u64,
+    headers_size: usize,
+}
+
+struct TreeStage {
+    metadata: TreeMeta,
+    meta_buf: Vec<u8>,
+    tree_buf: Vec<u8>,
+}
+
+pub(super) struct Serializer<'a> {
+    headers_btree: &'a InMemoryIndex,
+    header_stage: Option<HeaderStage>,
+    leaves_stage: Option<LeavesStage>,
+    tree_stage: Option<TreeStage>,
+}
+
+impl<'a> Serializer<'a> {
+    pub(super) fn new(headers_btree: &'a InMemoryIndex) -> Self {
+        Self {
+            headers_btree,
+            header_stage: None,
+            leaves_stage: None,
+            tree_stage: None,
+        }
+    }
+
+    pub(super) fn build(self) -> Result<(IndexHeader, TreeMeta, Vec<u8>)> {
+        let HeaderStage {
+            header,
+            filter_buf,
+            key_size: _,
+        } = self
+            .header_stage
+            .ok_or_else(|| anyhow!("Previous stages were not done"))?;
+        let LeavesStage {
+            leaves_buf,
+            leaves_offset: _,
+            leaf_size: _,
+            headers_size,
+        } = self
+            .leaves_stage
+            .ok_or_else(|| anyhow!("Previous stages were not done"))?;
+        let TreeStage {
+            metadata,
+            tree_buf,
+            meta_buf,
+        } = self
+            .tree_stage
+            .ok_or_else(|| anyhow!("Previous stages were not done"))?;
+
+        let hs = header.serialized_size()? as usize;
+        let fsize = header.filter_buf_size;
+        let msize = meta_buf.len();
+        let data_size = hs + fsize + headers_size + msize + leaves_buf.len() + tree_buf.len();
+        let mut buf = Vec::with_capacity(data_size);
+        serialize_into(&mut buf, &header)?;
+        buf.extend_from_slice(&filter_buf);
+        Self::append_headers(self.headers_btree, &mut buf)?;
+        buf.extend_from_slice(&meta_buf);
```

Comment:
Does it make sense to extend once for all three appends to avoid possible multiple allocations?

Reply:
So you suggest to chain filter_buf, meta_buf... with iterator and extend from it  for example? I am sure, that there wouldn't be redundant allocations because I allocate buffer with exact needed size (check line 67)

Reply:
Missed it - my bad 🤦‍♂️